Reword tasks relating to ServiceAccounts (#33654)
kubernetes/website: content/en/docs/reference/access-authn-authz/service-accounts-admin.md

  for specific tasks on demand. Separating ServiceAccount creation from the steps to
  onboard human users makes it easier for workloads to following the principle of
  least privilege.
- Auditing considerations for humans and service accounts may differ; the separation
  makes that easier to achieve.
- A configuration bundle for a complex system may include definition of various service
  accounts for components of that system. Because service accounts can be created
  without many constraints and have namespaced names, such configuration is
  usually portable.

## Bound service account token volume mechanism {#bound-service-account-token-volume}

{{< feature-state for_k8s_version="v1.22" state="stable" >}}

By default, the Kubernetes control plane (specifically, the
[ServiceAccount admission controller](#service-account-admission-controller))
adds a [projected volume](/docs/concepts/storage/projected-volumes/) to Pods,
and this volume includes a token for Kubernetes API access.

Here's an example of how that looks for a launched Pod:

```yaml
...
  - name: kube-api-access-<random-suffix>
    projected:

>       defaultMode: 420 # decimal equivalent of octal 0644

Reviewer: I would elide this line, the default mode depends on other fields in the pod

            path: token
        - configMap:
            items:
              - key: ca.crt
                path: ca.crt
            name: kube-root-ca.crt
        - downwardAPI:
            items:
              - fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.namespace
                path: namespace
```

That manifest snippet defines a projected volume that consists of three sources. In this case,
each source also represents a single path within that volume. The three sources are:

1. A `serviceAccountToken` source, that contains a token that the kubelet acquires from kube-apiserver
   The kubelet fetches time-bound tokens using the TokenRequest API. A token served for a TokenRequest expires
   either when the pod is deleted or after a defined lifespan (by default, that is 1 hour).
   The token is bound to the specific Pod and has the kube-apiserver as its audience.
   This mechanism superseded an earlier mechanism that added a volume based on a Secret,
   where the Secret represented the ServiceAccount for the Pod, but did not expire.
1. A `configMap` source. The ConfigMap contains a bundle of certificate authority data. Pods can use these

>    certificates to make sure that they are connecting to your cluster's kube-apiserver (and not to middlebox

Author: https://en.wikipedia.org/wiki/Middlebox - that's the term I'd use

it does the following when a Pod is created or modified:

1. If the pod does not have a `.spec.serviceAccountName` set, the admission controller sets the name of the
   ServiceAccount for this incoming Pod to `default`.
1. The admission controller ensures that the ServiceAccount referenced by the incoming Pod exists. If there
   is no ServiceAccount with a matching name, the admission controller rejects the incoming Pod. That check
   applies even for the `default` ServiceAccount.
1. Provided that neither the ServiceAccount's `automountServiceAccountToken` field nor the
   Pod's `automountServiceAccountToken` field is set to `false`:
   - the admission controller mutates the incoming Pod, adding an extra
     {{< glossary_tooltip text="volume" term_id="volume" >}} that contains
     a token for API access.
   - the admission controller adds a `volumeSource` to each container in the Pod.
     For Linux containers, that `volumeSource` is mounted at `/var/run/secrets/kubernetes.io/serviceaccount`
   - If the spec of the incoming Pod does already contain any `imagePullSecrets`, then the
     admission controller adds `imagePullSecrets`, copying them from the `ServiceAccount`.

> ### TokenRequest API

Reviewer 2: Would it be worth mentioning the new `kubectl create token` feature here as that uses the TokenRequest API and could be used by end-users to generate new tokens?

Author: Maybe. Is that a “makes the page better” change, or a “page doesn't make sense without it”. If it's the former, I'd prefer to move this one forward towards merge, and also track an issue about also documenting `kubectl create token` - maybe with a new task page.

Author: How's this new wording?

using the `--service-account-key-file` flag. The public key will be used to
verify the tokens during authentication.

### ServiceAccount admission controller

The modification of pods is implemented via a plugin
called an [Admission Controller](/docs/reference/access-authn-authz/admission-controllers/).
It is part of the API server.
This admission controller acts synchronously to modify pods as they are created or

> updated. When this plugin is active (and it is by default on most distributions), then

Reviewer: "as they are created." (no changes on pod update are permitted)

> it does the following when a Pod is created or modified:

Reviewer: ```suggestion
it does the following when a Pod is created:
```

1. If the pod does not have a `.spec.serviceAccountName` set, the admission controller sets the name of the
   ServiceAccount for this incoming Pod to `default`.
1. The admission controller ensures that the ServiceAccount referenced by the incoming Pod exists. If there
   is no ServiceAccount with a matching name, the admission controller rejects the incoming Pod. That check
   applies even for the `default` ServiceAccount.
1. Provided that neither the ServiceAccount's `automountServiceAccountToken` field nor the
   Pod's `automountServiceAccountToken` field is set to `false`:
   - the admission controller mutates the incoming Pod, adding an extra
     {{< glossary_tooltip text="volume" term_id="volume" >}} that contains
     a token for API access.

>    - the admission controller adds a `volumeSource` to each container in the Pod.

Reviewer: ```suggestion
   - the admission controller adds a `volumeMount` to each container in the Pod.
```

>      For Linux containers, that `volumeSource` is mounted at `/var/run/secrets/kubernetes.io/serviceaccount`

Reviewer: I'm not sure where windows surfaces a mount path of `/var/run/secrets/kubernetes.io/serviceaccount`, but the mount path injected is the same regardless of OS

Reviewer: this is skipped for any containers which already have a volumeMount at `/var/run/secrets/kubernetes.io/serviceaccount`

>      For Linux containers, that `volumeSource` is mounted at `/var/run/secrets/kubernetes.io/serviceaccount`

Reviewer: ```suggestion
     For Linux containers, that `volumeMount` is mounted at `/var/run/secrets/kubernetes.io/serviceaccount`
```

1. If the spec of the incoming Pod does already contain any `imagePullSecrets`, then the
   admission controller adds `imagePullSecrets`, copying them from the `ServiceAccount`.

### TokenRequest API

{{< feature-state for_k8s_version="v1.22" state="stable" >}}

You use the [TokenRequest](/docs/reference/kubernetes-api/authentication-resources/token-request-v1/)
subresource of a ServiceAccount to obtain a time-bound token for that ServiceAccount. Usually you don't

> need to call this API: the kubelet sets this up for you using a _projected volume_.

Reviewer: I would clarify what "usually" means... e.g. "You don't need to call this to obtain an API token for use within a container, since the kubelet sets this up for you using a _projected volume_."

If you want to use the TokenRequest API from `kubectl`, see
[Manually create an API token for a ServiceAccount](/docs/tasks/configure-pod-container/configure-service-account/#manually-create-an-api-token-for-a-serviceaccount).

The Kubernetes control plane (specifically, the ServiceAccount admission controller)
adds a projected volume to Pods, and the kubelet ensures that this volume contains a token
that lets containers authenticate as the right ServiceAccount.

(This mechanism superseded an earlier mechanism that added a volume based on a Secret,
where the Secret represented the ServiceAccount for the Pod but did not expire.)

Here's an example of how that looks for a launched Pod:

```yaml
...
  - name: kube-api-access-<random-suffix>
    projected:
      defaultMode: 420 # decimal equivalent of octal 0644
      sources:
        - serviceAccountToken:

>             expirationSeconds: 3597

Reviewer: `3597` does not match reality, either elide this line or keep the previous content of `3607`

        - downwardAPI:
            items:
              - fieldRef:
                  apiVersion: v1
                  fieldPath: metadata.namespace
                path: namespace
```

That manifest snippet defines a projected volume that combines information from three sources:

1. A `serviceAccountToken` source, that contains a token that the kubelet acquires from kube-apiserver
   The kubelet fetches time-bound tokens using the TokenRequest API. A token served for a TokenRequest expires
   either when the pod is deleted or after a defined lifespan (by default, that is 1 hour).
   The token is bound to the specific Pod and has the kube-apiserver as its audience.
1. A `configMap` source. The ConfigMap contains a bundle of certificate authority data. Pods can use these
   certificates to make sure that they are connecting to your cluster's kube-apiserver (and not to middlebox
   or an accidentally misconfigured peer).
1. A `downwardAPI` source. This `downwardAPI` volume makes the name of the namespace container the Pod available
   to application code running inside the Pod.

Any container within the Pod that mounts this volume can access the above information.

> See more details about [projected volumes](/docs/tasks/configure-pod-container/configure-projected-volume-storage/).

Reviewer: this content is confusing to include in the TokenRequest API section... it looks like a repeat of the content above, and is unrelated to direct use of the TokenRequest API

## Create additional API tokens {#create-token}

> To create the first, or subsequent, long lived API token for a ServiceAccount, create a

Reviewer: ```suggestion
To create a non-expiring Secret-based API token for a ServiceAccount, create a
```

Reviewer: should we preface this with a note or warning about the downsides of Secret-based tokens and recommend TokenRequest / `create token` instead when possible?

  namespace: examplens
  resourceVersion: "777"
  selfLink: /api/v1/namespaces/examplens/serviceaccounts/example-automated-thing
  uid: f23fd170-66f2-4697-b049-e1e266b7f835
secrets:
- name: example-automated-thing-token-zyxwv
```
Then, delete the Secret you now know the name of:
```shell
kubectl -n examplens delete secret/example-automated-thing-token-zyxwv
```

The control plane spots that the ServiceAccount is missing its Secret,
and creates a replacement:

```shell
kubectl -n examplens get serviceaccount/example-automated-thing -o yaml
```
```yaml
apiVersion: v1
kind: ServiceAccount
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"ServiceAccount","metadata":{"annotations":{},"name":"example-automated-thing","namespace":"examplens"}}
  creationTimestamp: "2019-07-21T07:07:07Z"
  name: example-automated-thing
  namespace: examplens
  resourceVersion: "1026"
  selfLink: /api/v1/namespaces/examplens/serviceaccounts/example-automated-thing
  uid: f23fd170-66f2-4697-b049-e1e266b7f835
secrets:
- name: example-automated-thing-token-4rdrh
```

You can see that there is now a new associated Secret with a different name. The

> old Secret is no longer valid.

Reviewer: since 1.24, a new secret is not auto-created and auto-added to the service account, so this line isn't really accurate
